x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheService.java

```diff
@@ -141,4 +202,129 @@ public void removeFromCache(final Predicate<CacheKey> predicate) {
     }
+
+    void setCacheSyncInterval(TimeValue interval) {
+        cacheSyncTask.setInterval(interval);
+    }
+
+    private void setMaxCacheFilesToSyncAtOnce(int maxCacheFilesToSyncAtOnce) {
+        this.maxCacheFilesToSyncAtOnce = maxCacheFilesToSyncAtOnce;
+    }
+
+    /**
+     * This method is invoked when a {@link CacheFile} notifies the current {@link CacheService} that it needs to be fsync on disk.
+     * <p>
+     * It adds the {@link CacheFile} instance to current set of cache files to synchronize.
+     *
+     * @param cacheFile the instance that needs to be fsync
+     */
+    void onCacheFileUpdate(CacheFile cacheFile) {
+        assert cacheFile != null;
+        cacheFilesToSync.offer(cacheFile);
+        numberOfCacheFilesToSync.incrementAndGet();
+    }
+
+    /**
+     * This method is invoked after a {@link CacheFile} is evicted from the cache.
+     * <p>
+     * It notifies the {@link CacheFile}'s eviction listeners that the instance is evicted.
+     *
+     * @param cacheFile the evicted instance
+     */
+    void onCacheFileRemoval(CacheFile cacheFile) {
+        IOUtils.closeWhileHandlingException(cacheFile::startEviction);
+    }
+
+    // used in tests
+    boolean isCacheFileToSync(CacheFile cacheFile) {
+        return cacheFilesToSync.contains(cacheFile);
+    }
+
+    /**
+     * Synchronize the cache files and their parent directories on disk.
+     *
+     * This method synchronizes the cache files that have been updated since the last time the method was invoked. To be able to do this,
+     * the cache files must notify the {@link CacheService} when they need to be fsync. When a {@link CacheFile} notifies the service the
+     * {@link CacheFile} instance is added to the current queue of cache files to synchronize referenced by {@link #cacheFilesToSync}.
+     *
+     * Cache files are serially synchronized using the {@link CacheFile#fsync()} method. When the {@link CacheFile#fsync()} call returns a
+     * non empty set of completed ranges this method also fsync the shard's snapshot cache directory, which is the parent directory of the
+     * cache entry. Note that cache files might be evicted during the synchronization.
+     */
+    protected void synchronizeCache() {
+        long count = 0L;
+        final Set<Path> cacheDirs = new HashSet<>();
+        final long startTimeNanos = threadPool.relativeTimeInNanos();
+        final long maxCacheFilesToSync = Math.min(numberOfCacheFilesToSync.get(), this.maxCacheFilesToSyncAtOnce);
+        for (long i = 0L; i < maxCacheFilesToSync; i++) {
+            if (lifecycleState() != Lifecycle.State.STARTED) {
+                logger.debug("stopping cache synchronization (cache service is closing)");
+                break;
+            }
+            final CacheFile cacheFile = cacheFilesToSync.poll();
+            if (cacheFile == null) {
+                logger.debug("stopping cache synchronization (no more cache files to fsync)");
```

Comment:
Could we theoretically assert that the service has been stopped here? (it seems the only way for the queue to be reduced is for this method to run and it should only ever run once at any point in time)

Reply:
Since this method is the only place where cache files are removed from the queue I think it makes more sense to `synchronized` the `synchronizeCache()` method and asserts here that we never poll a null `cacheFile`